Cannot move out of borrowed content in for loop

Hi there, 

I am probably not posting this in the right place but I seek help using this library. 
I am trying to iterate over a list of x509Name's using the following code in my template : 

```rust
{% for entry in certificate.subject_name().entries() %}

{% endfor %}
```

`certificate` being an instance of [X509](https://docs.rs/openssl/0.10.11/x86_64-unknown-linux-gnu/openssl/x509/struct.X509.html). I have the following error : `cannot move out of borrowed content`. I am pretty new to rust, this must be something easy but I have not yet full comprehension of the concept of ownership..

Just wanted to follow-up on this issue with a similar problem. Given the following structs:

```rust
pub struct Coursework {
    pub instructor: String,
    pub name: String,
    pub semester: String,
    pub short: String,
}

#[derive(Debug, Template)]
#[template(path = "markdown/courses.md", print = "code")]
pub struct MarkdownCoursesTemplate {
    pub coursework: Vec<Coursework>,
}
```

The generated code is

```rust
impl ::askama::Template for MarkdownCoursesTemplate {
    fn render_into(&self, writer: &mut ::std::fmt::Write) -> ::askama::Result<()> {
        writer.write_str("## Coursework\n\n<table class=\"table table-hover\">")?;
        writer.write_str("\n")?;
        for (_loop_index, course) in (&self.coursework).into_iter().enumerate() {
            writer.write_str("\n")?;
            writer.write_str("<tr>\n  <td class=\'col-md-1\'>")?;
            write!(writer, "{}", &{ course.semester })?;
            writer.write_str("</td>\n  <td><strong>")?;
            write!(writer, "{}", &{ course.name })?;
            writer.write_str("</strong> (")?;
            write!(writer, "{}", &{ course.short })?;
            writer.write_str("),")?;
            writer.write_str(" ")?;
            write!(writer, "{}", &{ course.instructor })?;
            writer.write_str("</td>\n</tr>")?;
            writer.write_str("\n")?;
        }
        writer.write_str("\n")?;
        writer.write_str("</table>")?;
        Ok(())
    }
    fn extension(&self) -> Option<&str> {
        Some("md")
    }
}
```

This fails compilation with 

```
error[E0507]: cannot move out of borrowed content
  --> src/markdown.rs:17:37
   |
17 |             write!(writer, "{}", &{ course.semester })?;
   |                                     ^^^^^^ cannot move out of borrowed content

error[E0507]: cannot move out of borrowed content
  --> src/markdown.rs:19:37
   |
19 |             write!(writer, "{}", &{ course.name })?;
   |                                     ^^^^^^ cannot move out of borrowed content

error[E0507]: cannot move out of borrowed content
  --> src/markdown.rs:21:37
   |
21 |             write!(writer, "{}", &{ course.short })?;
   |                                     ^^^^^^ cannot move out of borrowed content

error[E0507]: cannot move out of borrowed content
  --> src/markdown.rs:24:37
   |
24 |             write!(writer, "{}", &{ course.instructor })?;
   |                                     ^^^^^^ cannot move out of borrowed content

error: aborting due to 4 previous errors
```

Had the generated code been

```rust
impl ::askama::Template for MarkdownCVTemplate {
    fn render_into(&self, writer: &mut ::std::fmt::Write) -> ::askama::Result<()> {
        writer.write_str("## Coursework\n\n<table class=\"table table-hover\">")?;
        writer.write_str("\n")?;
        for (_loop_index, course) in (&self.coursework).into_iter().enumerate() {
            writer.write_str("\n")?;
            writer.write_str("<tr>\n  <td class=\'col-md-1\'>")?;
            write!(writer, "{}", &course.semester)?; // <--
            writer.write_str("</td>\n  <td><strong>")?;
            write!(writer, "{}", &course.name)?; // <--
            writer.write_str("</strong> (")?;
            write!(writer, "{}", &course.short)?; // <--
            writer.write_str("),")?;
            writer.write_str(" ")?;
            write!(writer, "{}", &course.instructor)?; // <--
            writer.write_str("</td>\n</tr>")?;
            writer.write_str("\n")?;
        }
        writer.write_str("\n")?;
        writer.write_str("</table>")?;
        Ok(())
    }
    fn extension(&self) -> Option<&str> {
        Some("md")
    }
}
```
it would have compiled. Is there a recommended workaround for this type of scenario?

--- a/askama_shared/src/generator.rs
+++ b/askama_shared/src/generator.rs
@@ -667,8 +667,26 @@ impl<'a, S: std::hash::BuildHasher> Generator<'a, S> {
                 ", _loop_item) in ::askama::helpers::TemplateLoop::new({}) {{",
                 expr_code
             )),
+            Expr::Array(..) => buf.writeln(&format!(
+                ", _loop_item) in ::askama::helpers::TemplateLoop::new({}.iter()) {{",
+                expr_code
+            )),
+            // If `iter` is a call then we assume it's something that returns
+            // an iterator. If not then the user can explicitly add the needed
+            // call without issues.
+            Expr::MethodCall(..) | Expr::PathCall(..) | Expr::Index(..) => buf.writeln(&format!(
+                ", _loop_item) in ::askama::helpers::TemplateLoop::new(({}).into_iter()) {{",
+                expr_code
+            )),
+            // If accessing `self` then it most likely needs to be
+            // borrowed, to prevent an attempt of moving.
+            _ if expr_code.starts_with("self.") => buf.writeln(&format!(
+                ", _loop_item) in ::askama::helpers::TemplateLoop::new(((&{}).into_iter())) {{",
+                expr_code
+            )),
+            // Otherwise, we borrow `iter` assuming that it implements `IntoIterator`.
             _ => buf.writeln(&format!(
-                ", _loop_item) in ::askama::helpers::TemplateLoop::new((&{}).into_iter()) {{",
+                ", _loop_item) in ::askama::helpers::TemplateLoop::new(({}).into_iter()) {{",
                 expr_code
             )),
         }?;

--- a/testing/tests/loops.rs
+++ b/testing/tests/loops.rs
@@ -67,3 +67,39 @@ fn test_for_range() {
         "foo (first)\nfoo (last)\nbar\nbar\nfoo\nbar\nbar\n"
     );
 }
+
+#[derive(Template)]
+#[template(source = "{% for i in [1, 2, 3] %}{{ i }}{% endfor %}", ext = "txt")]
+struct ForArrayTemplate {}
+
+#[test]
+fn test_for_array() {
+    let t = ForArrayTemplate {};
+    assert_eq!(t.render().unwrap(), "123");
+}
+
+#[derive(Template)]
+#[template(
+    source = "{% for i in [1, 2, 3].iter() %}{{ i }}{% endfor %}",
+    ext = "txt"
+)]
+struct ForMethodCallTemplate {}
+
+#[test]
+fn test_for_method_call() {
+    let t = ForMethodCallTemplate {};
+    assert_eq!(t.render().unwrap(), "123");
+}
+
+#[derive(Template)]
+#[template(
+    source = "{% for i in [1, 2, 3, 4, 5][3..] %}{{ i }}{% endfor %}",
+    ext = "txt"
+)]
+struct ForIndexTemplate {}
+
+#[test]
+fn test_for_index() {
+    let t = ForIndexTemplate {};
+    assert_eq!(t.render().unwrap(), "45");
+}

